From 10f3c3cebf2fb8dd8f9a2d72dfeda24dca7f2b2a Mon Sep 17 00:00:00 2001
From: Dylan Hung <dylan_hung@aspeedtech.com>
Date: Mon, 18 Jul 2022 13:56:42 +0800
Subject: [PATCH 13/30] i3c: aspeed: add API to set the static address in slave
 mode

When the I3C controller runs in slave mode, it is necessary for the
slave application to set the new static address according to different
user scenario in run time.  For example, there are two AST1030 BIC
boards with the same firmware attached on the same bus, it is needed to
set different static addresses for these two BIC boards to avoid address
confliction.  The firmware developer can refer to the sample code below
to set the new address:

```
&i3c1 {
	...
	// the static address of i3c1 was 0x9
	i3c1_smq:i3c-slave-mqueue@9 {
		...
		reg = <0x9>;
	};
};

const struct device *slave_controller;
uint8_t new_addr = 0x10

slave_controller =
device_get_binding(DT_BUS_LABEL(DT_NODELABEL(i3c1_smq)));

ret = i3c_slave_set_static_addr(slave_controller, new_addr);
```

then the static address of i3c1 will be updated to 0x10.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: Ibeb698f76a90a3c4bce3b0ae2d766b8f7603ea85
---
 drivers/i3c/i3c_aspeed.c  | 15 +++++++++++++++
 include/drivers/i3c/i3c.h |  9 +++++++++
 2 files changed, 24 insertions(+)

diff --git a/drivers/i3c/i3c_aspeed.c b/drivers/i3c/i3c_aspeed.c
index dd09a9724c..b3bb9fc9e6 100644
--- a/drivers/i3c/i3c_aspeed.c
+++ b/drivers/i3c/i3c_aspeed.c
@@ -1429,6 +1429,21 @@ int i3c_aspeed_slave_wait_data_consume(const struct device *dev)
 	return 0;
 }
 
+int i3c_aspeed_slave_set_static_addr(const struct device *dev, uint8_t static_addr)
+{
+	struct i3c_aspeed_config *config = DEV_CFG(dev);
+	struct i3c_register_s *i3c_register = config->base;
+	union i3c_device_addr_s device_addr;
+
+	config->assigned_addr = static_addr;
+
+	device_addr.value = i3c_register->device_addr.value;
+	device_addr.fields.static_addr = static_addr;
+	i3c_register->device_addr.value = device_addr.value;
+
+	return 0;
+}
+
 int i3c_aspeed_slave_get_dynamic_addr(const struct device *dev, uint8_t *dynamic_addr)
 {
 	struct i3c_aspeed_config *config = DEV_CFG(dev);
diff --git a/include/drivers/i3c/i3c.h b/include/drivers/i3c/i3c.h
index 48b4634bda..d258439b5b 100644
--- a/include/drivers/i3c/i3c.h
+++ b/include/drivers/i3c/i3c.h
@@ -164,6 +164,14 @@ int i3c_aspeed_master_request_ibi(struct i3c_dev_desc *i3cdev, struct i3c_ibi_ca
 int i3c_aspeed_master_enable_ibi(struct i3c_dev_desc *i3cdev);
 int i3c_aspeed_slave_register(const struct device *dev, struct i3c_slave_setup *slave_data);
 
+/**
+ * @brief set the static address of the i3c controller in slave mode
+ * @param dev the I3C controller in slave mode
+ * @param static_addr the new static address
+ * @return 0 if the static address is set
+ */
+int i3c_aspeed_slave_set_static_addr(const struct device *dev, uint8_t static_addr);
+
 /**
  * @brief get the assigned dynamic address of the i3c controller
  * @param dev the I3C controller in slave mode
@@ -235,6 +243,7 @@ int i3c_master_send_getbcr(const struct device *master, uint8_t addr, uint8_t *b
 #define i3c_master_request_ibi		i3c_aspeed_master_request_ibi
 #define i3c_master_enable_ibi		i3c_aspeed_master_enable_ibi
 #define i3c_slave_register		i3c_aspeed_slave_register
+#define i3c_slave_set_static_addr	i3c_aspeed_slave_set_static_addr
 #define i3c_slave_send_sir		i3c_aspeed_slave_send_sir
 #define i3c_slave_put_read_data		i3c_aspeed_slave_put_read_data
 #define i3c_slave_wait_data_consume	i3c_aspeed_slave_wait_data_consume
-- 
2.24.1

